File: solutions/ruby/2022/02/lib/rock_paper_scissors/part_1.rb
# frozen_string_literal: true

require_relative '../rock_paper_scissors'

module RockPaperScissors
  module Part1
    class PlayerOrOpponent
      attr_reader :shape

      def self.from_shape_input(shape_input)
        new shapes_by_input[shape_input]
      end

      def initialize(shape)
        @shape = shape
      end
    end

    class Opponent < PlayerOrOpponent
      SHAPES_BY_INPUT = { 'A' => Shape::ROCK, 'B' => Shape::PAPER, 'C' => Shape::SCISSORS }.freeze

      def self.shapes_by_input
        SHAPES_BY_INPUT
      end
    end

    class Player < PlayerOrOpponent
      SHAPES_BY_INPUT = { 'X' => Shape::ROCK, 'Y' => Shape::PAPER, 'Z' => Shape::SCISSORS }.freeze

      def self.shapes_by_input
        SHAPES_BY_INPUT
      end
    end

    class Round < Round
      def self.from_input_line(input_line)
        opponent_shape_input, player_shape_input = input_line.split
        new Opponent.from_shape_input(opponent_shape_input), Player.from_shape_input(player_shape_input)
      end

      def shape_score
        SHAPE_SCORES[player.shape]
      end

      def outcome
        return Outcome::DRAW if opponent.shape == player.shape

        OUTCOMES[[opponent.shape, player.shape]]
      end
    end

    class Game < Game
      def self.round_class
        Round
      end
    end
  end
end

File: solutions/rust/2022/02/src/rock_paper_scissors/part_1.rs
use std::str::FromStr;

use super::{Opponent, Outcome, Shape};

#[derive(Debug)]
struct Player {
    shape: Shape,
}

#[derive(Debug)]
struct Round {
    opponent: Opponent,
    player: Player,
}

#[derive(Debug)]
pub struct Game {
    rounds: Vec<Round>,
}

impl FromStr for Player {
    type Err = String;

    fn from_str(shape_input: &str) -> Result<Self, Self::Err> {
        let shape = match shape_input {
            "X" => Shape::Rock,
            "Y" => Shape::Paper,
            "Z" => Shape::Scissors,
            _ => return Err(format!("{:?} is not a valid shape", shape_input).to_string()),
        };
        Ok(Player { shape })
    }
}

impl FromStr for Round {
    type Err = String;

    fn from_str(round_input: &str) -> Result<Self, Self::Err> {
        let mut slices = round_input.split(' ');

        let opponent = slices.next().map(Opponent::from_str);
        let player = slices.next().map(Player::from_str);

        if let (Some(Ok(opponent)), Some(Ok(player)), None) = (opponent, player, slices.next()) {
            Ok(Round { opponent, player })
        } else {
            Err(format!("{:?} is not a valid round", round_input).to_string())
        }
    }
}

impl FromStr for Game {
    type Err = String;

    fn from_str(input: &str) -> Result<Self, Self::Err> {
        let maybe_rounds = input
            .trim_end()
            .split("\n")
            .map(|input_line| Round::from_str(input_line))
            .collect::<Result<Vec<_>, _>>();

        if let Ok(rounds) = maybe_rounds {
            Ok(Game { rounds })
        } else {
            Err(format!("{:?} are not valid rounds", maybe_rounds).to_string())
        }
    }
}

impl Round {
    fn shape_score(&self) -> u32 {
        match &self.player.shape {
            Shape::Rock => 1,
            Shape::Paper => 2,
            Shape::Scissors => 3,
        }
    }

    fn outcome(&self) -> Outcome {
        match (&self.opponent.shape, &self.player.shape) {
            (Shape::Scissors, Shape::Rock) => Outcome::Win,
            (Shape::Paper, Shape::Scissors) => Outcome::Win,
            (Shape::Rock, Shape::Paper) => Outcome::Win,
            (Shape::Rock, Shape::Scissors) => Outcome::Loose,
            (Shape::Scissors, Shape::Paper) => Outcome::Loose,
            (Shape::Paper, Shape::Rock) => Outcome::Loose,
            _ => Outcome::Draw,
        }
    }

    fn outcome_score(&self) -> u32 {
        match &self.outcome() {
            Outcome::Win => 6,
            Outcome::Draw => 3,
            Outcome::Loose => 0,
        }
    }

    fn score(&self) -> u32 {
        &self.shape_score() + &self.outcome_score()
    }
}

impl Game {
    pub fn score(&self) -> u32 {
        self.rounds.iter().map(|round| round.score()).sum()
    }
}
